criu/arch/aarch64/crtools.c

```diff
@@ -20,10 +21,87 @@
 #include "cpu.h"
 #include "restorer.h"
 #include "compel/infect.h"
+#include "pstree.h"
+
+extern unsigned long getauxval(unsigned long type);
 
 #define assign_reg(dst, src, e) dst->e = (__typeof__(dst->e))(src)->e
 
-int save_task_regs(void *x, user_regs_struct_t *regs, user_fpregs_struct_t *fpsimd)
+static int save_pac_keys(int pid, CoreEntry *core)
+{
+	struct user_pac_address_keys paca;
+	struct user_pac_generic_keys pacg;
+	PacKeys *pac_entry;
+	long pac_enabled_key;
+	struct iovec iov;
+	int ret;
+
+	unsigned long hwcaps = getauxval(AT_HWCAP);
+
+	hwcaps = 0;
+
+	pr_debug("%d: Dumping PAC keys\n", pid);
+	pac_entry = xmalloc(sizeof(PacKeys));
+	if (!pac_entry)
+		return -1;
+	core->ti_aarch64->pac_keys = pac_entry;
+	pac_keys__init(pac_entry);
+
+	if (hwcaps & HWCAP_PACA) {
+		PacAddressKeys *pac_address_keys;
+
+		iov.iov_base = &paca;
+		iov.iov_len = sizeof(paca);
+		if ((ret = ptrace(PTRACE_GETREGSET, pid, NT_ARM_PACA_KEYS, &iov))) {
+			pr_perror("Failed to obtain FPU registers for %d", pid);
+			return -1;
+		}
+		pac_address_keys = xmalloc(sizeof(PacAddressKeys));
+		if (!pac_address_keys)
+			return -1;
+		pac_address_keys__init(pac_address_keys);
+		pac_entry->pac_address_keys = pac_address_keys;
+		pac_address_keys->apiakey_lo = paca.apiakey;
+		pac_address_keys->apiakey_hi = paca.apiakey >> 64;
+		pac_address_keys->apibkey_lo = paca.apibkey;
+		pac_address_keys->apibkey_hi = paca.apibkey >> 64;
+		pac_address_keys->apdakey_lo = paca.apdakey;
+		pac_address_keys->apdakey_hi = paca.apdakey >> 64;
+		pac_address_keys->apdbkey_lo = paca.apdbkey;
+		pac_address_keys->apdbkey_hi = paca.apdbkey >> 64;
+
+		iov.iov_base = &pac_enabled_key;
+		iov.iov_len = sizeof(pac_enabled_key);
+		ret = ptrace(PTRACE_GETREGSET, pid, NT_ARM_PAC_ENABLED_KEYS, &iov);
+		if (ret) {
+			pr_perror("Failed to obtain FPU registers for %d", pid);
+			return -1;
+		}
+
+		pac_address_keys->pac_enabled_key = pac_enabled_key;
+
+	}
+	if (hwcaps & HWCAP_PACG) {
+		PacGenericKeys *pac_generic_keys;
+
+		iov.iov_base = &pacg;
+		iov.iov_len = sizeof(pacg);
+		if ((ret = ptrace(PTRACE_GETREGSET, pid, NT_ARM_PACG_KEYS, &iov))) {
```

Comment:
Do we need to worry about the kernels that are not compiled with PAC support (`CONFIG_ARM64_PTR_AUTH`)?  This allows the branch to be entered (as the hardware exhibits the capability), but the `ptrace` call will fail.

Reply:
HWCAP_PACG isn't reported if the kernel isn't compiled with PAC.